flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -954,6 +968,7 @@ public void heartbeatFromResourceManager(final ResourceID resourceID) {
 		}
 		return checkpointCoordinator
 			.triggerSavepoint(System.currentTimeMillis(), targetDirectory)
+			.thenApply(CheckpointExecuteResult::getCompletedCheckpoint)
```

Comment:
If there was a savepoint failure, this call on the `CheckpointExecuteResult` will trigger an illegal state exception and even lose the trace of the original exception. I dont think that is what we want here, but handling two different paths?

Reply:
accept